File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
n = int(input())
dp = [0] * (n + 1)
step = [0] * (n + 1)

for i in range(1, n + 1):
    step[i] = int(input())

if n == 1:
    print(step[1])
    exit()
if n == 2:
    print(sum(step[:3]))
    exit()

dp[1] = step[1]
dp[2] = step[1] + step[2]

for i in range(3, n + 1):
    dp[i] = max(step[i] + dp[i - 2], step[i] + step[i - 1] + dp[i - 3])

print(dp[n])

File: 백준/Silver/2579. 계단 오르기/계단 오르기.kt
import kotlin.math.max

fun main() {
    val br = System.`in`.bufferedReader()
    val bw = System.`out`.bufferedWriter()

    val input = br.readLine().toInt()
    val dp = IntArray(input + 1) { 0 }

    val step = mutableListOf<Int>()
    step.add(0)
    repeat(input) {
        step.add(br.readLine().toInt())
    }

    when (input) {
        0 -> {
            bw.write("${step[0]}")
        }

        1 -> {
            bw.write("${step[1]}")
        }

        2 -> {
            bw.write("${step[1] + step[2]}")
        }

        else -> {
            dp[1] = step[1]
            dp[2] = step[1] + step[2]

            for (i in 3..input) {
                dp[i] = max(step[i] + dp[i - 2], step[i] + step[i - 1] + dp[i - 3])
            }
            bw.write("${dp[input]}")
        }
    }

    bw.close()
}